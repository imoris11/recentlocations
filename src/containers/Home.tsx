import {fetchUser} from '../actions/usersActions';
import {fetchLocation, clearAllLocations, removeLocation} from '../actions/locationActions';
import {BottomTabNavigationProp} from '@react-navigation/bottom-tabs';
import {RouteProp} from '@react-navigation/native';
import React, {Component} from 'react';
import {
  StyleSheet,
  Text,
  View,
  FlatList,
  TouchableOpacity,
  PermissionsAndroid,
  Platform,
} from 'react-native';
import {connect} from 'react-redux';
import {AppTabParamList} from '../App';
import {User, SingleItem, Location} from '../types';
import SingleListItem from '../components/ListItem';
import Geolocation, {
  GeolocationResponse,
  GeolocationError,
} from '@react-native-community/geolocation';
import {convertTimestampToDate} from '../helpers';
import {CurrentLocation} from '../components/CurrentLocation';

export interface HomeProps {
  fetchUser: typeof fetchUser;
  user: User;
  locations: Location[];
  navigation: BottomTabNavigationProp<AppTabParamList, 'Home'>;
  route: RouteProp<AppTabParamList, 'Home'>;
  fetchLocation: typeof fetchLocation;
  clearAllLocations: typeof clearAllLocations;
  removeLocation: typeof removeLocation;
}

export class Home extends Component<HomeProps> {
  state = {
    watchId: 0,
  };
  componentDidMount() {
    this.props.fetchUser('1');
    this.getLocation();
  }

  requestLocationPermission = async () => {
    try {
      const granted = await PermissionsAndroid.request(
        PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
        {
          title: 'Allow WeHaul to use your location',
          message:
            'WeHaul needs to use your location for trip tracking ' +
            'so you can keep the customer updated on your progress.',
          buttonNegative: 'Cancel',
          buttonPositive: 'OK',
        },
      );
      if (granted === PermissionsAndroid.RESULTS.GRANTED) {
        return true;
      } else {
        return false;
      }
    } catch (err) {
      console.warn(err);
      return false;
    }
  };

  getLocation = () => {
    const accessGranted = Platform.OS === 'ios' ? true : this.requestLocationPermission();
    if (accessGranted) {
      const watchId = Geolocation.watchPosition(
        this.distanceChangedSuccessfully,
        this.distanceChangeFailed,
        {enableHighAccuracy: true, distanceFilter: 100},
      );
      this.setState({watchId});
    }
  };

  distanceChangedSuccessfully = (position: GeolocationResponse) => {
    const coords = `${position.coords.latitude},${position.coords.longitude}`;
    this.props.fetchLocation(coords);
  };

  distanceChangeFailed = (error: GeolocationError) => {
    //do something eith error?
    console.log(error);
  };

  componentWillUnmount() {
    Geolocation.clearWatch(this.state.watchId);
  }

  removeListItem = (payload: number) => {
    this.props.removeLocation(payload);
  };

  _renderItem = ({item}: SingleItem) => {
    return (
      <SingleListItem
        key={item.timestamp}
        removeListItem={(timeStamp: number) => this.removeListItem(timeStamp)}
        item={item}
      />
    );
  };

  clearStore = () => {
    this.props.clearAllLocations();
  };

  _keyExtractor = (item: Location) => {
    return `${item.timestamp}`;
  };

  render() {
    const {locations, user} = this.props;
    const recentLocation = locations[0] || [];
    const timestamp = convertTimestampToDate(recentLocation.timestamp);
    return (
      <View style={styles.container}>
        <View style={styles.content}>
          <View style={styles.headerContainer}>
            <Text style={styles.headerText}>Location Manager</Text>
          </View>
          <CurrentLocation
            timestamp={timestamp || ''}
            formattedAddress={recentLocation.formattedAddress || 'Enable Location Services'}
            username={user?.name}
          />
          <View style={styles.listContainer}>
            <Text style={styles.sectionHeader}>Previous Locations</Text>
            <FlatList
              keyExtractor={(item) => `${item.timestamp}`}
              data={locations}
              renderItem={this._renderItem}
            />
          </View>
          <TouchableOpacity style={styles.clearAllContainer} onPress={() => this.clearStore()}>
            <Text style={styles.clearAllText}>Clear All</Text>
          </TouchableOpacity>
        </View>
      </View>
    );
  }
}

const mapStateToProps = (state: any) => ({
  user: state.users.user,
  locations: state.locations.locations,
});

const mapDispatchToProps = {
  fetchUser,
  fetchLocation,
  clearAllLocations,
  removeLocation,
};

export default connect(mapStateToProps, mapDispatchToProps)(Home);

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  content: {
    marginTop: 30,
    paddingHorizontal: 20,
    flex: 1,
  },
  headerContainer: {
    marginVertical: 20,
  },
  headerText: {
    fontSize: 18,
    fontWeight: '700',
  },
  sectionHeader: {
    color: 'grey',
    fontWeight: '600',
  },
  clearAllContainer: {
    backgroundColor: '#0e64ef',
    position: 'absolute',
    bottom: 20,
    justifyContent: 'center',
    alignItems: 'center',
    marginHorizontal: 20,
    left: 20,
    right: 20,
    borderRadius: 3,
    padding: 15,
  },
  clearAllText: {
    color: '#FFF',
    fontWeight: '500',
  },
  listContainer: {
    flex: 1,
  },
});
