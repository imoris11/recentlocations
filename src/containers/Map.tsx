import React, {Component} from 'react';
import {StyleSheet} from 'react-native';
import MapView, {Marker} from 'react-native-maps';
import {connect} from 'react-redux';
import {AppTabParamList} from '../App';
import {BottomTabNavigationProp} from '@react-navigation/bottom-tabs';
import {Location} from '../types';
import {RouteProp} from '@react-navigation/native';
import {EmptyMessage} from '../components/EmptyMessage';

export interface HomeProps {
  locations: Location[];
  navigation: BottomTabNavigationProp<AppTabParamList, 'Home'>;
  route: RouteProp<AppTabParamList, 'Home'>;
}

export class Map extends Component<HomeProps> {
  render() {
    const {locations} = this.props;
    return locations.length ? (
      <MapView
        style={styles.container}
        showsUserLocation={true}
        followsUserLocation={true}
        showsMyLocationButton={true}
        initialRegion={{
          latitude: locations[0]?.lat,
          longitude: locations[0]?.lng,
          latitudeDelta: 0.0922,
          longitudeDelta: 0.0421,
        }}>
        {locations.map((location: Location, index) => (
          <Marker
            key={location.timestamp + index}
            coordinate={{latitude: location.lat, longitude: location.lng}}
            title={'Recently visited'}
            description={location.formattedAddress}
          />
        ))}
      </MapView>
    ) : (
      <EmptyMessage />
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});

const mapStateToProps = (state: any) => {
  return {
    locations: state.locations.locations,
  };
};

export default connect(mapStateToProps, null)(Map);
