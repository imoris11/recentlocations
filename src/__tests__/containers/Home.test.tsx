import {Home} from '../../containers/Home';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import {render, fireEvent} from 'react-native-testing-library';

Enzyme.configure({adapter: new Adapter()});

const createTestProps = (props?: object) => ({
  fetchUser: jest.fn(),
  user: {},
  ...props,
});

jest.mock('react-redux', () => {
  return {
    connect: jest.fn().mockReturnValue(() => jest.fn()),
  };
});
jest.mock('../../actions/usersActions', () => {
  return {
    fetchUser: jest.fn().mockReturnValue('mock user action'),
  };
});

jest.mock(
  'react-native-localization',
  () =>
    class RNLocalization {
      language = 'en';
      props: any;

      constructor(props: any) {
        this.props = props;
        this.setLanguage(this.language);
      }

      setLanguage(interfaceLanguage: any) {
        this.language = interfaceLanguage;
        if (this.props[interfaceLanguage]) {
          const localizedStrings = this.props[this.language];
          for (const key in localizedStrings) {
            if (localizedStrings.hasOwnProperty(key)) {
              (this as any)[key] = localizedStrings[key];
            }
          }
        }
      }
    },
);

describe('Home', () => {
  const fetchUser = jest.fn();
  const navigate = jest.fn();
  const props: any = createTestProps({
    fetchUser,
    navigation: {
      navigate,
    },
    locations: [],
  });
  const {toJSON} = render(<Home {...props} />);
  it('should call fetchUser', () => {
    expect(fetchUser).toBeCalled();
  });

  it('should match snapshot', () => {
    expect(toJSON()).toMatchSnapshot();
  });

  it('clears store when clear all is clicked', () => {
    const MOCK_FN = jest.fn();
    const {getByText} = render(<Home {...props} clearAllLocations={MOCK_FN} />);
    const element = getByText('Clear All');
    fireEvent(element, 'click');
    expect(MOCK_FN).toHaveBeenCalled();
  });
});

describe('ConnectedHome', () => {
  const mockConnect = require('react-redux').connect;
  const mapStateToProps = mockConnect.mock.calls[0][0];
  it('should map user from state to props', () => {
    const user = {id: '1'};
    const mockState = {users: {user}, locations: {}};
    const props = mapStateToProps(mockState);
    expect(props.user).toEqual(user);
  });
});
