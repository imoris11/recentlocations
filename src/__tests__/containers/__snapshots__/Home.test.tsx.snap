// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "marginTop": 30,
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      style={
        Object {
          "marginVertical": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 18,
            "fontWeight": "700",
          }
        }
      >
        Location Manager
      </Text>
    </View>
    <View
      style={
        Object {
          "marginVertical": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "grey",
            "fontWeight": "600",
          }
        }
      >
        Current Location
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginVertical": 15,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#ee6723",
              "borderRadius": 20,
              "height": 40,
              "justifyContent": "center",
              "width": 40,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FFF",
                "fontWeight": "800",
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "marginLeft": 10,
              "marginRight": 40,
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Object {
                "fontSize": 16,
                "fontWeight": "500",
              }
            }
          >
            Enable Location Services
          </Text>
          <Text
            style={
              Object {
                "color": "grey",
                "marginTop": 2,
              }
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "grey",
            "fontWeight": "600",
          }
        }
      >
        Previous Locations
      </Text>
      <RCTScrollView
        data={Array []}
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View />
      </RCTScrollView>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#0e64ef",
          "borderRadius": 3,
          "bottom": 20,
          "justifyContent": "center",
          "left": 20,
          "marginHorizontal": 20,
          "opacity": 1,
          "padding": 15,
          "position": "absolute",
          "right": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFF",
            "fontWeight": "500",
          }
        }
      >
        Clear All
      </Text>
    </View>
  </View>
</View>
`;
